### PR TITLE
refactor: config registration

### PR DESCRIPTION
**Pull request by itayd**
- config printing and registration so it wont be necessary to actually initialize the service to populate all configurations for printing. 
- add `Validate()` method to all config to validate without actually starting stuff up.

the underlying motivation here is that the next step I'd like also to document each configuration param in code (tbd how, either field tag or a custom method?).

### CODE REVIEW

File: internal/backend/auth/authloginhttpsvc/config.go [hunk 1]
```diff
@@ -28,20 +50,44 @@ type descopeConfig struct {
...
+func (c descopeConfig) Validate() error {
+	if !c.Enabled {
+		return nil
+	}
+
+	if c.ProjectID == "" {
+		return errors.New("project_id is required")
+	}
+
+	if c.ManagementKey == "" {
```
Comment (daabr) [inline internal/backend/auth/authloginhttpsvc/config.go]: AFAIK management key isn't required. For example, I have a Descope config in my local server that doesn't have it.

File: internal/backend/auth/authloginhttpsvc/config.go [hunk 2]
```diff
@@ -28,20 +50,44 @@ type descopeConfig struct {
...
 type Config struct {
 	GoogleOAuth oauth2Config  `koanf:"google_oauth"`
 	GithubOAuth oauth2Config  `konf:"github_oauth"`
 	Descope     descopeConfig `koanf:"descope"`
 
 	// Allowed login patterns, separated by commas.
 	// Pattern format is either of:
-	// - "*"       - matches any login
+	// - "" or "*" - matches any login
 	// - "*@host"  - matches any login from host
```
Comment (daabr) [inline internal/backend/auth/authloginhttpsvc/config.go]:
```suggestion
	// - "*@domain"  - matches any login, but only from domain
```

**Review by daabr — COMMENTED** [review]
Didn't finish my review, but here are a couple of comments

File: backend/svc/config.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package svc
+
+import (
+	"go.autokitteh.dev/autokitteh/internal/backend/config"
+)
+
+type Config = config.Config
```
Comment (tebeka) [inline backend/svc/config.go:7]: I don't like the liberal use we do with type aliases, they are intended for code refactoring.

File: internal/backend/config/set.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
+package config
+
+import (
+	"fmt"
+
+	"go.autokitteh.dev/autokitteh/sdk/sdkerrors"
+)
+
+type ComponentConfig interface {
```
Comment (tebeka) [inline internal/backend/config/set.go:9]: Rename to `Validated` or `Validator`

File: internal/backend/config/set.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+type emptyComponentConfig struct{}
+
+func (emptyComponentConfig) Validate() error { return nil }
+
+var EmptyComponentConfig emptyComponentConfig
+
+type baseSet interface {
+	hasDefault() bool
+	default_() ComponentConfig
```
Comment (tebeka) [inline internal/backend/config/set.go]: Not happy with `default_` and in general lower case, unexported, functions in an interface.

File: internal/backend/config/set.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+	hasDev() bool
+	dev() ComponentConfig
+
+	hasTest() bool
+	test() ComponentConfig
+}
+
+type Set[T ComponentConfig] struct {
```
Comment (tebeka) [inline internal/backend/config/set.go:23]: IMO we can get rid of `Set` and have only one configuration. The settings of `dev`, `test` and `default` should be set from the outside.

File: internal/backend/config/set.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+func (s Set[T]) dev() ComponentConfig      { return *s.Dev }
+func (s Set[T]) test() ComponentConfig     { return *s.Test }
+
+func (s Set[T]) hasDefault() bool { return s.Default != nil }
+func (s Set[T]) hasDev() bool     { return s.Dev != nil }
+func (s Set[T]) hasTest() bool    { return s.Test != nil }
+
+var EmptySet = Set[emptyComponentConfig]{Default: &EmptyComponentConfig}
+
+func (set *Set[T]) Choose(mode Mode) (chosen T, err error) {
```
Comment (tebeka) [inline internal/backend/config/set.go:31]: Why do we need this and not access `s.Default` etc?

File: internal/backend/secrets/config.go [hunk 8]
```diff
@@ -17,11 +20,17 @@ type Config struct {
...
 }
 
-var (
-	Configs = configset.Set[Config]{
-		Default: &Config{},
-		Dev: &Config{
-			Provider: SecretProviderDatabase,
-		},
+func (c Config) Validate() error {
+	if c.Provider != "" && !kittehs.ContainedIn(providers...)(c.Provider) {
```
Comment (tebeka) [inline internal/backend/secrets/config.go]: Use built-in `slices.Contains`

File: internal/backend/store/store.go [hunk 9]
```diff
@@ -23,7 +24,16 @@ type Config struct {
 	ServerURL string `koanf:"server_url"`
 }
 
-var Configs = configset.Set[Config]{
+func (c Config) Validate() error {
+	if c.ServerURL == "" || c.ServerURL == miniredisURL {
+		return nil
+	}
+
+	_, err := url.Parse(c.ServerURL)
```
Comment (tebeka) [inline internal/backend/store/store.go:32]: Try to connect as well?

**Review by tebeka — APPROVED** [review]
Approving for the added validation. Still think code is too complex.

File: internal/backend/auth/authloginhttpsvc/config.go (hunk 1 above)
Comment (itayd) [reply internal/backend/auth/authloginhttpsvc/config.go]: removed mgmt key all togethor.

File: internal/backend/auth/authloginhttpsvc/config.go (hunk 2 above)
Comment (itayd) [reply internal/backend/auth/authloginhttpsvc/config.go]: done

File: backend/svc/config.go (hunk 3 above)
Comment (itayd) [reply backend/svc/config.go:7]: this is only for exposing internal stuff through this specific package. I don't think I am using aliases anywhere else tbh

File: internal/backend/config/set.go (hunk 4 above)
Comment (itayd) [reply internal/backend/config/set.go:9]: the intention is to add more functionality in the future, so `Validator` is too narrow

File: internal/backend/config/set.go (hunk 5 above)
Comment (itayd) [reply internal/backend/config/set.go]: removed all functions, now only constraining known sets.

File: internal/backend/config/set.go (hunk 6 above)
Comment (itayd) [reply internal/backend/config/set.go:23]: don't think so - we want to be able in the future, or even very soon, to list all values and not just keys per configuration mode.

File: internal/backend/config/set.go (hunk 7 above)
Comment (itayd) [reply internal/backend/config/set.go:31]: it's choosing the right one depending on mode - not sure i understand the q.

File: internal/backend/secrets/config.go (hunk 8 above)
Comment (itayd) [reply internal/backend/secrets/config.go]: oooohhhh nice

File: internal/backend/store/store.go (hunk 9 above)
Comment (itayd) [reply internal/backend/store/store.go:32]: nah, that actually has a side effect.

**Pull request merged**